core/controllers/voice_artist.py

```diff
@@ -126,3 +145,92 @@ def post(self, unused_exploration_id):
         user_services.record_user_started_state_translation_tutorial(
             self.user_id)
         self.render_json({})
+
+
+class UserVoicoverApplicationsHandler(base.BaseHandler):
+    """Handler for the voiceover applications submitted by the user."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    @acl_decorators.can_access_voiceover_applications
+    def get(self, purpose):
+        """Handles GET requests."""
+        if purpose == feconf.VOICEOVER_APPLICATION_REVIEW:
+            voiceover_applications = (
+                voiceover_services.get_reviewable_voiceover_applications(
+                    self.user_id))
+        elif purpose == feconf.VOICEOVER_APPLICATION_STATUS:
+            status = self.request.get('status')
+            voiceover_applications = (
+                voiceover_services.get_user_submitted_voiceover_applications(
+                    self.user_id, status=status))
+        else:
+            raise self.PageNotFoundException
+
+        self.values = {
+            'voiceover_applications': [
+                v_a.to_dict() for v_a in voiceover_applications]
+        }
+        self.render_json(self.values)
+
+
+class VoicoverApplicationHandler(base.BaseHandler):
+    """Handler for the voiceover application."""
+
+    @acl_decorators.can_review_voiceover_application
+    def put(self, voiceover_application_id):
+
+        action = self.payload.get('action')
+
+        if action == suggestion_models.ACTION_TYPE_ACCEPT:
+            voiceover_services.accept_voiceover_application(
+                voiceover_application_id, self.user_id)
+        elif action == suggestion_models.ACTION_TYPE_REJECT:
+            review_message = self.payload.get('review_message')
+            voiceover_services.reject_voiceover_application(
+                voiceover_application_id, self.user_id, review_message)
+        else:
+            raise self.InvalidInputException('Invalid action.')
+
+        self.render_json({})
+
+    @acl_decorators.can_submit_voiceover_application
+    def post(self):
+        raw_audio_file = self.request.get('raw_audio_file')
+
+        target_type = self.payload.get('target_type')
+        target_id = self.payload.get('target_id')
+        language_code = self.payload.get('language_code')
+        voiceover_content = self.payload.get('voiceover_content')
+        filename = (
+            self.user_id + utils.generate_random_string(6) + '.mp3')
+
+        try:
+            _save_audio_file(
+                raw_audio_file, filename,
+                feconf.ENTITY_TYPE_VOICEOVER_APPLICATION, target_id,
+                self.user_id)
+            voiceover_services.create_new_voiceover_application(
```

Comment:
This doesn't need to be part of the try/catch block right? Let's try to keep the try catch blocks as small as possible.

Reply:
I see this pattern in other handlers as well, I think we cannot throw 400* errors from service layer so we do it this way in the controller? (not sure!)

https://github.com/oppia/oppia/blob/8933f81c8ce7880e46aca30f621f364202e7a704/core/controllers/editor.py#L128-L143


Note: `create_new_voiceover_application` will throw validation errors so should only catch validationErrors?

Reply:
Hmmm, yeah if we know what errors we want to catch, it is better. It becomes harder to debug errors if there is a lot of stuff between the try/catch.

Reply:
Done!